tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Logging/NLog/LogsInjection/DiagnosticContextHelper.cs

```diff
@@ -0,0 +1,136 @@
+﻿// <copyright file="DiagnosticContextHelper.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using Datadog.Trace.DuckTyping;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.Logging.NLog.LogsInjection
+{
+    internal static class DiagnosticContextHelper
+    {
+        public static MappedDiagnosticsLogicalContextSetterProxy GetMdlcProxy(Assembly nlogAssembly)
+        {
+            var mdlcType = nlogAssembly.GetType("NLog.MappedDiagnosticsLogicalContext");
+            if (mdlcType is not null)
+            {
+                // NLog 4.3+
+                var createTypeResult = DuckType.GetOrCreateProxyType(typeof(MappedDiagnosticsLogicalContextSetterProxy), mdlcType);
+                if (createTypeResult.Success)
+                {
+                    // NLog 4.6+
```

Comment:
I don't understand these 4.3+ and 4.6+ comments. Are they in the right place?

Reply:
The 4.6+ comment is in the right place, because we'll enter the if block if the `NLog.MappedDiagnosticsLogicalContext` class has the `SetScoped` method. You can see that the same check is done for LibLog here: https://github.com/DataDog/dd-trace-dotnet/pull/2096/files/03ba4ca3af25ae440ba5e42ffa7885a7f43b63bd#diff-df9d635c71ca6fa3fcd3a1d61b9f351c0ffd5918d1505ba9195529e4bd0584d3L40

Let me look into the 4.3+ comment a little more...

Reply:
For the 4.3+ comment, I believe that should be updated to 4.1 since that is when the type is introduced into the library: https://github.com/NLog/NLog/commit/5821834a206544cee6312f3e236d0872349e37b2#diff-4434139d866e82cd90c857ffde0cc307c33b0fd36cd3e6ee1e4419a5519317df / https://www.nuget.org/packages/NLog/4.1.0

Or if Andrew has more context, perhaps he can clarify when he gets back. Either way, I'd like to not upset the CI gods and merge this as is and then add clarifying comments in my Logs Injection Fix PR

Reply:
> For the 4.3+ comment, I believe that should be updated to 4.1

I think that's correct. Much of this code is copy-pasted from types I used with direct log submission, where we need to do even more fiddling as 4.0/4.1/4.3/5.6 are all different.

Also worth noting that `MappedDiagnosticsLogicalContextSetterProxy` is obsolete in NLog 5, so at some point we probably need to special case that one too so we don't break in future versions